pkg/policy/rule.go

```diff
@@ -41,26 +42,70 @@ func (r *rule) validate() error {
 	return nil
 }
 
-func mergeL4Port(ctx *SearchContext, r api.PortRule, p api.PortProtocol, dir string, proto string, resMap L4PolicyMap) int {
-	fmt := p.Port + "/" + proto
-	v, ok := resMap[fmt]
+func adjustL4PolicyIfNeeded(fromEndpoints []api.EndpointSelector, policy *L4Filter) bool {
+
+	if len(policy.FromEndpoints) == 0 && len(fromEndpoints) > 0 {
+		logrus.Debugf("skipping this L4 policy since it is already covered by an existing one.")
+		return true
+	}
+
+	if len(policy.FromEndpoints) > 0 && len(fromEndpoints) == 0 {
+		// new policy is more permissive than the existing one
+		// use a more permissive one
+		policy.FromEndpoints = nil
+	}
+	return false
+}
+
+func mergeL4Port(fromEndpoints []api.EndpointSelector, r api.PortRule, p api.PortProtocol,
+	dir string, proto string, resMap L4PolicyMap) int {
+
+	portProto := p.Port + "/" + proto
+	v, ok := resMap[portProto]
 	if !ok {
-		resMap[fmt] = CreateL4Filter(r, p, dir, proto)
+		resMap[portProto] = CreateL4Filter(fromEndpoints, r, p, dir, proto)
 		return 1
 	}
-	l4Filter := CreateL4Filter(r, p, dir, proto)
+
+	if adjustL4PolicyIfNeeded(fromEndpoints, &v) && len(r.Rules.HTTP) == 0 {
+		// skip this policy as it is already covered and it does not contain l7 rules
+		return 0
+	}
+
+	l4Filter := CreateL4Filter(fromEndpoints, r, p, dir, proto)
 	if l4Filter.L7Parser != "" {
 		v.L7Parser = l4Filter.L7Parser
 	}
 	if l4Filter.L7RedirectPort != 0 {
 		v.L7RedirectPort = l4Filter.L7RedirectPort
 	}
-	v.L7Rules = append(v.L7Rules, l4Filter.L7Rules...)
-	resMap[fmt] = v
+
+	// if (1) the existing *ingress* rule did not have a wildcard EP
+	// AND (2) the new rule does not have explicit fromEndpoints
+	// THEN we need to copy all existing l7 rules to the wildcard ep
+	if _, ok := v.L7RulesPerEp[WildcardEndpointSelector]; l4Filter.Ingress && !ok && len(fromEndpoints) == 0 {
+		v.L7RulesPerEp[WildcardEndpointSelector] = L7Rules{}
+		wildcardEp := v.L7RulesPerEp[WildcardEndpointSelector]
+		for _, existingL7Rules := range v.L7RulesPerEp {
+			wildcardEp = append(wildcardEp, existingL7Rules...)
+		}
+	}
+
+	for hash, newL7Rules := range l4Filter.L7RulesPerEp {
+		if ep, ok := v.L7RulesPerEp[hash]; ok {
+			ep = append(ep, newL7Rules...)
```

Comment:
This doesn't work as you're expecting, replace with
`v.L7RulesPerEp[hash] = append(ep, newL7Rules...)`